ROS2 Pose publisher

sandra-Robotnik opened the pull request.

sandra-Robotnik committed:
ROS2 structure

sandra-Robotnik committed:
Initial ROS2 node

sandra-Robotnik committed:
Added a timer

sandra-Robotnik committed:
Changed ros2 tf dependencies

sandra-Robotnik committed:
Created the publisher and the parameters

sandra-Robotnik committed:
ROS2 Tf transformation

rafa-martin commented on package.xml:
```diff
@@ -1,24 +1,28 @@
-  <version>0.1.3</version>
+  <version>0.0.0</version>
   <description>This package provides the current position and orientation of the robot in the map with PoseWithCovarianceStamped format.</description>
-
```
Why removed the original author? should be the same as previous?

rafa-martin commented on src/pose_publisher.cpp:
```diff
@@ -1,85 +1,130 @@
-/*
- * Copyright (c) 2014, Zhi Yan <zhi.yan@mines-douai.fr>
- * All rights reserved.
- * Redistribution and use in source and binary forms, with or without
- * modification, are permitted provided that the following conditions are met:
- *
- *     * Redistributions of source code must retain the above copyright
- *       notice, this list of conditions and the following disclaimer.
- *     * Redistributions in binary form must reproduce the above copyright
- *       notice, this list of conditions and the following disclaimer in the
- *       documentation and/or other materials provided with the distribution.
- *     * Neither the name of the University of California, Berkeley nor the
- *       names of its contributors may be used to endorse or promote products
- *       derived from this software without specific prior written permission.
- *
- * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS "AS IS" AND ANY
- * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
- * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
- * DISCLAIMED. IN NO EVENT SHALL THE REGENTS AND CONTRIBUTORS BE LIABLE FOR ANY
- * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
- * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
- * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
- * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
- * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
- * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- */
```
Needs to keep the original license and the author of the code.

rafa-martin commented on src/pose_publisher.cpp:
```diff
@@ -1,85 +1,130 @@
+private:
+    void timer_callback()
+    {
+        RCLCPP_INFO(get_logger(),  "In timer_callback");
```
Is it a debug info? should be throttle info?

rafa-martin commented on src/pose_publisher.cpp:
```diff
@@ -1,85 +1,130 @@
+    {
+        RCLCPP_INFO(get_logger(),  "In timer_callback");
+
+        std::string fromFrameRel = (tf_prefix + "/" + base_frame).c_str();
+        std::string toFrameRel = (tf_prefix + "/" + map_frame).c_str();
```
const variable, no changes after assign it in this line

rafa-martin commented on src/pose_publisher.cpp:
```diff
@@ -1,85 +1,130 @@
+        // std::string fromFrameRel = (tf_prefix + "/" + map_frame).c_str();
+        // std::string toFrameRel = (tf_prefix + "/" + base_frame).c_str();
+
+        RCLCPP_INFO(get_logger(),  "fromFrameRel topic (target_frame): %s", fromFrameRel.c_str());
+        RCLCPP_INFO(get_logger(),  "toFrameRel topic (source_frame): %s",toFrameRel.c_str());
```
Is it a debug info? should be throttle info?

rafa-martin commented on src/pose_publisher.cpp:
```diff
@@ -1,85 +1,130 @@
+                toFrameRel, fromFrameRel,
+                tf2::TimePointZero);
+        } catch (const tf2::TransformException & ex) {
+            RCLCPP_INFO(
```
Is it a warn info?

rafa-martin commented on src/pose_publisher.cpp:
```diff
@@ -1,85 +1,130 @@
+        try {
+            tf_transform = tf_buffer_->lookupTransform(
+                toFrameRel, fromFrameRel,
+                tf2::TimePointZero);
```
If I remember well, this use latest available transform from buffer. Should be used current timestamp of wall timer?

https://github.com/ros2/geometry2/blob/ad72a82961e82470d2c9848840dd6e8eee44b127/tf2_ros/include/tf2_ros/buffer.h#L137

rafa-martin commented on src/pose_publisher.cpp:
```diff
@@ -1,85 +1,130 @@
 
-      pose_publisher.publish(pose_stamped);
+            geometry_msgs::msg::PoseWithCovarianceStamped pose_stamped;
+            pose_stamped.header.stamp = rclcpp::Clock{}.now();
```
should be used the original timepoint from ``lookupTransform` method. which is given `geometry_msgs::msg::TransformStamped`

rafa-martin commented on src/pose_publisher.cpp:
```diff
@@ -1,85 +1,130 @@
-  }
+    /* Here it is created the shared pointers to the Publisher and timer objects
+    * defined above, and it is also created the variable count*/
+    rclcpp::TimerBase::SharedPtr timer_{nullptr};
+    std::string tf_prefix = "robot";
+    std::string map_frame = "map";
+    std::string base_frame = "base_link";
+    std::string topic_republish = "pose";
+    rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr pose_publisher_{nullptr};
```
Please, keep same format, `{}` initialization or `=`

rafa-martin reviewed (changes requested):
Small tips

sandra-Robotnik committed:
Updated package.xml file

sandra-Robotnik committed:
Fixed ros2 pose publisher node

sandra-Robotnik committed:
Passed the timer frequency rate as a parameter

sandra-Robotnik committed:
Created launch file

sandra-Robotnik committed:
Changed params name to match with the launch file

sandra-Robotnik committed:
Updated changelog

sandra-Robotnik committed:
Added the proper QoS
 - QoS profile:
   -> History (Depth): KEEP_LAST (1)
   -> Reliability: RELIABLE
   -> Durability: TRANSIENT_LOCAL

sandra-Robotnik committed:
Updated changelog file

sandra-Robotnik committed:
Release candidate 0.1.4-rc02

sandra-Robotnik committed:
Release 0.1.4
